### ISSUE TITLE
Performance optimization for DifferentiationInterfaceJAX.jl

### ISSUE THREAD

**Issue by gdalle** [pabloferz/DLPack.jl#42]
Hi @pabloferz!

Following your kind invitation, here's an example use of https://github.com/gdalle/DifferentiationInterfaceJAX.jl where most of the overhead is caused by array conversion. I'm curious if there is any faster way to do things?
You can take a look at the code inside my repo, it is very straightforward (15 LOCs for the gradient in total).

```julia
using BenchmarkTools
using DifferentiationInterface
using DifferentiationInterfaceJAX  # to install from URL
using PythonCall  # add JAX to your CondaPkg environment

jax = pyimport("jax")
jnp = pyimport("jax.numpy")

backend = AutoJAX()

f(x) = jnp.sum(jnp.square(x))

xj = rand(Float32, 10^5);
xp = pytensor(xj);

gradient(f, backend, xj) == jltensor(jax.grad(f)(xp))  # true
```

Benchmark results:

```julia
julia> @btime gradient(f, $prep, $backend, $xj);
  134.050 μs (55 allocations: 4.04 KiB)

julia> @btime $(jax.jit(jax.grad(f)))($xp);
  4.024 μs (2 allocations: 32 bytes)
```

Profiling flame graph:

![image](https://github.com/user-attachments/assets/940a4e6f-efaa-49e3-82b2-1a68f99e2a86)

**Comment by gdalle** [comment]
Here's a benchmark for the other half of the overhead: moving back from JAX tensors to Julia arrays:
```julia
using BenchmarkTools
using DLPack: share, from_dlpack
using PythonCall

jax = pyimport("jax")
jnp = pyimport("jax.numpy")

fp(xp) = jnp.square(xp)

function fj(xp)
    yp = fp(xp)
    yj = from_dlpack(yp)
    return yj
end

xj = Float32.(1:10^5)
xp = share(xj, jax.dlpack.from_dlpack)
```

```julia
julia> @btime fp($xp);
  37.098 μs (11 allocations: 184 bytes)

julia> @btime fj($xp);
  49.526 μs (30 allocations: 2.02 KiB)
```
